cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -776,6 +817,161 @@ protected void doGetSchedules(HttpResponder responder, ApplicationId application
+   * </code></pre>
+   * </p><p>
+   * The response will be an array of JsonObjects each of which will contain the three input parameters
+   * as well as a "schedules" field, which is a list of {@link ScheduledRuntime} object.
+   * </p><p>
+   * If an error occurs in the input (for the example above, App1 does not exist), then all JsonObjects for which the
+   * parameters have a valid status will have the status field but all JsonObjects for which the parameters do not have
+   * a valid status will have an error message and statusCode.
+   */
+  @POST
+  @Path("/nextruntime")
+  public void batchNextRunTimes(FullHttpRequest request,
+                                HttpResponder responder,
+                                @PathParam("namespace-id") String namespaceId) throws Exception {
+    List<BatchProgram> batchPrograms = validateAndGetBatchInput(request, BATCH_PROGRAMS_TYPE);
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(batchRunTimes(namespaceId, batchPrograms, false)));
+  }
+
+  /**
+   * Fetches scheduled run times for a set of programs.
+   *
+   * @param namespace namespace of the programs
+   * @param programs the list of programs to fetch scheduled run times
+   * @param previous {@code true} to get the previous scheduled times; {@code false} to get the next scheduled times
+   * @return a list of {@link BatchProgramSchedule} containing the result
+   * @throws SchedulerException if failed to fetch schedules
+   */
+  private List<BatchProgramSchedule> batchRunTimes(String namespace, Collection<? extends BatchProgram> programs,
+                                                   boolean previous) throws Exception {
+    List<ProgramId> programIds = programs.stream()
+      .map(p -> new ProgramId(namespace, p.getAppId(), p.getProgramType(), p.getProgramId()))
+      .collect(Collectors.toList());
+
+    Set<ApplicationId> appIds = programIds.stream().map(ProgramId::getParent).collect(Collectors.toSet());
+    Map<ApplicationId, ApplicationSpecification> appSpecs = store.getApplications(appIds);
+
+    List<BatchProgramSchedule> result = new ArrayList<>();
+    for (ProgramId programId : programIds) {
+      ApplicationSpecification spec = appSpecs.get(programId.getParent());
+      if (spec == null) {
+        result.add(new BatchProgramSchedule(programId, HttpResponseStatus.NOT_FOUND.code(),
+                                            new NotFoundException(programId.getParent()).getMessage(), null));
+        continue;
+      }
+      try {
+        Store.ensureProgramExists(programId, spec);
+        result.add(new BatchProgramSchedule(programId, HttpResponseStatus.OK.code(), null,
+                                            getScheduledRunTimes(programId, previous)));
+      } catch (NotFoundException e) {
+        result.add(new BatchProgramSchedule(programId, HttpResponseStatus.NOT_FOUND.code(), e.getMessage(), null));
+      } catch (BadRequestException e) {
```

Comment:
doesn't look like Store.ensureProgramExists throws this, should it be throwing NotFound instead of IllegalArgument?

Reply:
oh this gets thrown from getScheduledRunTimes. But it seems like IllegalArgument should still be fixed in ensureProgramExists()

Reply:
The IllegalArgumentException should never happen though, since the appSpec was coming from fetching the app of the program id. The implementation of the ensureProgramExists is more to guard against misuse (bug) in the call, which can't be raised in normal case.